tests/jdom/newtester.ts

```diff
@@ -0,0 +1,168 @@
+import { JDBus } from "../../src/jdom/bus"
+import { DEVICE_ANNOUNCE } from "../../src/jdom/constants"
+import { JDDevice } from "../../src/jdom/device"
+import JDServiceProvider from "../../src/jdom/serviceprovider"
+import JDServiceServer from "../../src/jdom/serviceserver"
+import { assert } from "../../src/jdom/utils"
+import {
+    DebugConsoleUi,
+    SynchronizationTimingOptions,
+    TestDriver,
+    TesterEvent,
+} from "../../src/tstester/base"
+import { BusTester } from "../../src/tstester/testwrappers"
+import { loadSpecifications } from "../testutils"
+import { FastForwardScheduler } from "./scheduler"
+import { CreatedServerService } from "./tester"
+
+// Set equals is not a built-in operation.
+function setEquals<T>(set1: Set<T>, set2: Set<T>): boolean {
+    if (set1.size != set2.size) {
+        return false
+    }
+    set1.forEach(value => {
+        if (!set2.has(value)) {
+            return false
+        }
+    })
+    return true
+}
+
+// A bus test wrapper that uses the fast-forward bus and provides test helper functions
+// to spin up virtual devices.
+// TODO: should this / can this be merged w/ FastForwardTestDriver?
+//   both use the FF scheduler, and it might make sense to combine the classes
+export class FastForwardBusTester extends BusTester {
+    // Wrapper that provides bus construction, initializaiton, and teardown
+    static async withTestBus(
+        test: (bus: FastForwardBusTester) => Promise<void>
+    ) {
+        const bus = new FastForwardBusTester()
+        bus.start()
+        await test(bus)
+        bus.stop()
+    }
+
+    readonly scheduler: FastForwardScheduler
+    // Unlike BusTester, we initialize a bus here so it uses the FF scheduler
+    constructor() {
+        super(
+            new JDBus([], {
+                scheduler: new FastForwardScheduler(),
+            })
+        )
+        loadSpecifications() // TODO this reimplements mkBus
+        assert(this.bus.scheduler instanceof FastForwardScheduler)
+        this.scheduler = this.bus.scheduler as FastForwardScheduler
+    }
+
+    // start() and stop() are made available, but recommend using withTestBus() instead
+    public start() {
+        this.bus.start()
+    }
+
+    public stop() {
+        this.bus.stop()
+    }
+
+    // Waits for all the devices (by deviceId) to be announced on the bus.
+    protected async waitForAnnounce(deviceIds: string[]) {
```

Comment:
handle empty array, noop

Reply:
Done, added a fast termination if array is empty